server/src/main/java/com/flipkart/varadhi/auth/DefaultAuthorizationProvider.java

```diff
@@ -0,0 +1,139 @@
+package com.flipkart.varadhi.auth;
+
+import com.flipkart.varadhi.entities.UserContext;
+import com.flipkart.varadhi.utils.YamlLoader;
+import io.vertx.core.Future;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
+
+
+@Slf4j
+public class DefaultAuthorizationProvider implements AuthorizationProvider {
+    private DefaultAuthorizationConfiguration configuration;
+    private volatile boolean initialised = false;
+
+    @Override
+    public Future<Boolean> init(AuthorizationOptions authorizationOptions) {
+        if (!this.initialised) {
+            synchronized (this) {
+                if (!this.initialised) {
+                    this.configuration =
+                            YamlLoader.loadConfig(authorizationOptions.getConfigFile(), DefaultAuthorizationConfiguration.class);
+                    this.initialised = true;
+                }
+            }
+        }
+        return Future.succeededFuture(true);
+    }
+
+    /**
+     * Checks if the user {@code UserContext} is authorized to perform action {@code ResourceAction} on resource path.
+     *
+     * @param userContext Information on the user, contains the user identifier
+     * @param action      Action being performed by the user which is to be authorized
+     * @param resource    Full schemaless URI of the resource on which the action is to be authorized.
+     *                    Must be of format: {org_id}/{team_id}/{project_id}/{topic|queue|subscription}
+     *
+     * @return {@code Future<Boolean>} a future result expressing True/False decision
+     */
+    @Override
+    public Future<Boolean> isAuthorized(UserContext userContext, ResourceAction action, String resource) {
+        List<Pair<ResourceType, String>> leafToRootResourceIds = resolveOrderedFromLeaf(action, resource);
+        var result = leafToRootResourceIds.stream().filter(pair -> StringUtils.isNotBlank(pair.getValue()))
+                .anyMatch(pair -> isAuthorizedInternal(userContext.getSubject(), action, pair.getValue()));
+        return Future.succeededFuture(result);
+    }
+
+    /**
+     * Parse the resource path based on the action and resolve the final resourceIDs for each resourceType.
+     * @param resourcePath uri of the resource
+     * @return List of pairs having resource type to its id. List is used so that we can impose ordering from leaf to root nodes.
+     */
+    private List<Pair<ResourceType, String>> resolveOrderedFromLeaf(ResourceAction action, String resourcePath) {
+        String[] segments = resourcePath.split("/");
+
+        // build the list in reverse order specified: ROOT -> ORG -> TEAM -> PROJECT -> TOPIC|SUBSCRIPTION|QUEUE
+        List<Pair<ResourceType, String>> resourceIdTuples = new ArrayList<>();
+        // handle leaf node case
+        if (isActionOnLeafNode(action)) {
```

Comment:
Not a special case, we want to know if there are any topic or subscription id provided in the resource path. Eg: Say in case of project create, there will be no topic or subscription id present so no leaf node is added in the list of tuples.